sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormatWriter.scala

```diff
@@ -314,21 +339,40 @@ object FileFormatWriter extends Logging {
 
           recordsInFile = 0
           releaseResources()
+          numOutputRows += recordsInFile
           newOutputWriter(fileCounter)
         }
 
         val internalRow = iter.next()
+        val startTime = System.nanoTime()
         currentWriter.write(internalRow)
+        timeOnCurrentFile += (System.nanoTime() - startTime)
```

Comment:
I'm neutral about the per-row time tracking. Comparing with actual data writing, it should be faster. IMHO, so the performance penalty may be ignored?

In both `SingleDirectoryWriteTask` and `DynamicPartitionWriteTask`, in each iteration, we first pull the row from the iterator and then write the row out (and measure the timing). So the time spent on pulling rows from data pipeline should be already excluded. That's why I'm not worried it may be inaccurate due to data pipelining.